### PR TITLE
Add Travis CI

### PR DESCRIPTION
**Pull request by fscheiner**
This adds automated builds on Travis CI (on `ppc64le` arch for now) and supports the following tasks (see [travis-ci/run_task_inside_docker.sh](https://github.com/fscheiner/UberFTP/blob/automated-builds/travis-ci/run_task_inside_docker.sh) for details) for now:
* `build` (as there are no tests defined for UberFTP I renamed the `tests` task to `build`) - (test-)builds UberFTP
* `tarballs`(should maybe renamed to `tarball`) - creates the source tarball (`make dist`) for UberFTP

...and since today also:
* `srpms`
* `rpms`

Check possible results on https://app.travis-ci.com/github/fscheiner/UberFTP/builds

****

Based on @matyasselmeci's Travis-CI scripts from the [GCT](https://github.com/gridcf/gct)

### CODE REVIEW

**Comment by fscheiner** [comment]
@msalle, @matyasselmeci, @ellert:
Before implementing the `srpms` task (needed for uploads of source tarballs to `repo.gridcf.org`) and GitHub Actions (which will actually perform the uploads) we need to decide on a few things:

- [x] 1. Should we adapt the `/packaging/[...]` directory structure of the GCT also for UberFTP? If yes, then we need to move the UberFTP RPM spec file there (i.e. `/packaging/fedora/uberftp.spec`)

Ok, I went for this "solution" now, see https://github.com/fscheiner/UberFTP/tree/automated-builds-continued, but I stumbled upon an issue I'll describe in a separate comment below.

UPDATE (2021-12-02): The following will be implemented in another PR.

~- [ ] 2. Where should we put the UberFTP source tarball(s), into https://repo.gridcf.org/uberftp/sources? As long as it's separate from the GCT, https://repo.gridcf.org/gct6/sources makes no sense IMO.~

**Review by msalle — APPROVED** [review]
I think I'm probably not the best to review this one. In any case, presuming it's mostly copy&paste from the rest of GCT, this seems all fine.
One thing:  `travis-ci/run_task_inside_docker.sh` has an almost completely outdated list of platforms: CentOS6, F26 and F27 are all long out of support. OTOH we probably want RockyLinux, Almalinux or CentOS Stream 8 and currently supported Fedoras.

**Comment by fscheiner** [comment]
@msalle 
> One thing: `travis-ci/run_task_inside_docker.sh` has an almost completely outdated list of platforms: CentOS6, F26 and F27 are all long out of support. OTOH we probably want RockyLinux, Almalinux or CentOS Stream 8 and currently supported Fedoras.

I ignored thatt because I didn't want to change too much at once. Indeed, RockyLinux and Almalinux are long due for testing. but I'm unsure if they provide any Docker images for ppc64le (or even arm64/aarch64). If someone has experience with these, please let us know.

AFAIK CentOS Stream 8 is not "identical" to - and maybe not even binary compatible with - RHEL8, so I don't see it as a target for our build tests (i.e. something that will replace our CentOS7 target in the future). I'd rather go for one or both of the earlier mentioned OSes as target as they want to be what CentOS once was.

**Comment by msalle** [comment]
> @msalle
> 
> > One thing: `travis-ci/run_task_inside_docker.sh` has an almost completely outdated list of platforms: CentOS6, F26 and F27 are all long out of support. OTOH we probably want RockyLinux, Almalinux or CentOS Stream 8 and currently supported Fedoras.
> 
> I ignored thatt because I didn't want to change too much at once. Indeed, RockyLinux and Almalinux are long due for testing. but I'm unsure if they provide any Docker images for ppc64le (or even arm64/aarch64). If someone has experience with these, please let us know.

Both RockyLinux  and AlmaLinux seem to have arm64, not ppc64le, in their repos and as docker images on docker hub. OTOH our reference platform is in any case amd64/x86_64.

> AFAIK CentOS Stream 8 is not "identical" to - and maybe not even binary compatible with - RHEL8, so I don't see it as a target for our build tests (i.e. something that will replace our CentOS7 target in the future). I'd rather go for one or both of the earlier mentioned OSes as target as they want to be what CentOS once was.

The point is that different sites will be using different distros. CERN is probably going for CentOS Stream, Desy AFAIK for AlmaLinux and several others big sites for RockyLinux.
 Building the GCT should be reasonably straightforward for all I guess.

**Comment by fscheiner** [comment]
> Both RockyLinux and AlmaLinux seem to have arm64, not ppc64le, in their repos and as docker images on docker hub. OTOH our reference platform is in any case amd64/x86_64.

Yes, but for Travis-CI we can't use them as long as we don't have build credits, ppc64le and arm64 are not charged by Travis-CI as the machinery is provided by manufacturers.

> The point is that different sites will be using different distros. CERN is probably going for CentOS Stream, Desy AFAIK for AlmaLinux and several others big sites for RockyLinux.

As long as it's compatible with each other and RHEL(8). Why CERN might want to use something different is out of my understanding.

@matyasselmeci 
What will OSG use in the future as replacement for CentOS7 and why?

**Comment by fscheiner** [comment]
> Both RockyLinux and AlmaLinux seem to have arm64, not ppc64le, in their repos and as docker images on docker hub.

Maybe IBM can convince the maintainers of RockyLinux and AlmaLinux to also support ppc64le? @gerrith3, @ezeeyahoo

**Comment by fscheiner** [comment]
@msalle 
About the currently supported Fedoras: I don't know for sure, if our existing RPM spec files work unchanged on Fedora - here I mean the ones from the GCT, not the one from UberFTP whose spec file is not overly complex, so should probably work anywhere. I believe they differ from the ones @ellert is using over at Fedora, but I didn't compare them recently.

I mean, if we want to have test builds for UberFTP on multiple OSes, we should also implement that for the GCT.

_...it was at this moment that he knew he had opened pandora's box..._ ;-)

**Comment by fscheiner** [comment]
**Please ignore, as outdated with the latest changes.**

****

Back to the original topic:
As per https://github.com/gridcf/UberFTP/pull/16#issuecomment-973091609 I have a problem when running `./make_rpms.sh`:

```
[johndoe@gridftp-5 travis-ci]$ ./make_rpms.sh
Wrote: /home/johndoe/git-projects/UberFTP/packaging/rpmbuild/SRPMS/uberftp-2.9-1.gct.src.rpm
*** Building uberftp ***
Error! Logs follow:
error: Failed build dependencies:
	globus-gssapi-gsi-devel is needed by uberftp-2.9-1.gct.x86_64
Installing /home/johndoe/git-projects/UberFTP/packaging/rpmbuild/SRPMS/uberftp-2.9-1.gct.src.rpm
Building target platforms: x86_64-linux
Building for target x86_64-linux
```

I don't want to make a PR for https://github.com/fscheiner/UberFTP/commit/beeb89e8786e7be4859c2174de601c26dcbae2d1 until it works, so want to solve that issue first.

@matyasselmeci 
In [make_rpms.sh on line 107](https://github.com/fscheiner/UberFTP/blob/60cce7774b6135d42344ea0d6d900bc4bf2cc2b9/travis-ci/make_rpms.sh#L107) I see that it removes all GCT packages (I extended that to also remove an installed UberFTP package) prior to the build:

```C
[...]
remove_installed_gct_rpms()
{
    rpm -qa | egrep '^globus-|^myproxy|^uberftp' \
        | xargs --no-run-if-empty \
            sudo rpm -e $pkgs_to_rm
}
[...]
```

But how can this work if a build depends on a `globus-*-devel` package?

I know that [`run_task_inside_docker.sh`](https://github.com/fscheiner/UberFTP/blob/60cce7774b6135d42344ea0d6d900bc4bf2cc2b9/travis-ci/run_task_inside_docker.sh) should actually pre-install some build dependencies - I deactivated that and install the UberFTP deps via the standard `packages` list (see [line 29](https://github.com/fscheiner/UberFTP/blob/60cce7774b6135d42344ea0d6d900bc4bf2cc2b9/travis-ci/run_task_inside_docker.sh#L29)) - but for example in the GCT this never installs any build requirements on `globus*` packages, though some spec files over at the GCT do require other `globus*` packages (e.g. [globus-gridftp-server.spec](https://github.com/gridcf/gct/blob/master/packaging/fedora/globus-gridftp-server.spec)). So I don't understand, why this could have worked for the GCT then. It doesn't work for UberFTP though.

UPDATE: Well, I just recognize that [the deploy step only builds SRPMs](https://github.com/gridcf/gct/blob/master/.github/workflows/build-test-deploy.yml#L30), so building RPMs would not be needed for the upload of the tarball to repo.gridcf.org, but still it should work, as this was also meant to allow users to builds RPMs locally.

UPDATE2: I think I have a possible solution to make `make_rpms.sh` work as intended, see [these changes](https://github.com/fscheiner/UberFTP/commit/beeb89e8786e7be4859c2174de601c26dcbae2d1#diff-ba7e92275f5f45949f964680a114d580c890e85e3c4668c8197ddc55422175c1) in https://github.com/fscheiner/UberFTP/commit/beeb89e8786e7be4859c2174de601c26dcbae2d1 for details.

**Comment by gerrith3** [comment]
> > Both RockyLinux and AlmaLinux seem to have arm64, not ppc64le, in their repos and as docker images on docker hub.
> 
> Maybe IBM can convince the maintainers of RockyLinux and AlmaLinux to also support ppc64le? @gerrith3, @ezeeyahoo

I will take a look but getting an entire distro supported can be difficult aka expensive.  It may depend on where they are pulling their sources from - if they are using fedora or equivalent build environments that are copied into RockyLinux or AlmaLinux there may be a chance.

**Comment by fscheiner** [comment]
@gerrith3 
> I will take a look but getting an entire distro supported can be difficult aka expensive. It may depend on where they are pulling their sources from - if they are using fedora or equivalent build environments that are copied into RockyLinux or AlmaLinux there may be a chance.

Great to hear that, I assume both RockyLinux and AlmaLinux are using the official sources of RHEL8 (which supports ppc64le), i.e. like CentOS8 did until Red Hat changed that to CentOS Stream 8.

**Comment by msalle** [comment]
> @msalle About the currently supported Fedoras: I don't know for sure, if our existing RPM spec files work unchanged on Fedora - here I mean the ones from the GCT, not the one from UberFTP whose spec file is not overly complex, so should probably work anywhere. I believe they differ from the ones @ellert is using over at Fedora, but I didn't compare them recently.
> 
> I mean, if we want to have test builds for UberFTP on multiple OSes, we should also implement that for the GCT.
> 
> _...it was at this moment that he knew he had opened pandora's box..._ ;-)

I think that for simplicity it would probably then be better to just remove those completely. Building (or at least appearing to build?) something on those long outdated platforms doesn't make sense to me. And then you can keep the box closed for now (-;

**Comment by fscheiner** [comment]
> I think that for simplicity it would probably then be better to just remove those completely. Building (or at least appearing to build?) something on those long outdated platforms doesn't make sense to me. And then you can keep the box closed for now (-;

I've done that with https://github.com/fscheiner/UberFTP/commit/beeb89e8786e7be4859c2174de601c26dcbae2d1 and will incorporate that here soon.

**Comment by fscheiner** [comment]
@msalle:
Ok, all additional changes are now included, A successful build is available on https://app.travis-ci.com/github/fscheiner/UberFTP/builds/242419144. I'll now re-request your review if you don't mind. The review requests for @matyasselmeci and @ellert are still open, so if they like they can also review the complete PR.

File: packaging/fedora/uberftp.spec [hunk 1]
```diff
@@ -8,7 +8,8 @@ Summary:        GridFTP-enabled ftp client
 
 License:        NCSA
 URL:            https://github.com/gridcf/UberFTP
-Source0:        https://github.com/gridcf/UberFTP/archive/%{gittag}/%{name}-%{version}.tar.gz
+#Source0:        https://github.com/gridcf/UberFTP/archive/%{gittag}/%{name}-%{version}.tar.gz
+Source0:        https://github.com/gridcf/UberFTP/archive/%{gittag}/uberftp-client-2.9-Beta.tar.gz
```
Comment (msalle) [inline packaging/fedora/uberftp.spec]: I'm a bit confused about this one?
Firstly where is the `-client` coming from (after all there is no server)? And is it wise to hardcode the name and version there?
Also it doesn't currently exist, is it produced as part of a still running build? (Sorry for my ignorance)

File: packaging/fedora/uberftp.spec [hunk 2]
```diff
@@ -20,7 +21,8 @@ It supports GSI authentication, parallel data channels and
 third party transfers.
 
 %prep
-%setup -q -n UberFTP-%{gittag}
+#%setup -q -n UberFTP-%{gittag}
+%setup -q -n uberftp-client-2.9-Beta
```
Comment (msalle) [inline packaging/fedora/uberftp.spec]: see other comment: why the client? And why hardcoded version here?

File: travis-ci/make_rpms.sh [hunk 3]
```diff
@@ -114,6 +114,9 @@ build_all_for_target()
             return 1
         fi
 
+        # install build requires as per https://stackoverflow.com/a/44443092
+        sudo yum install -y $( rpmbuild ${nocheck:+--nocheck} --target $target --rebuild "$rpmpath" 2>&1 | fgrep 'is needed by' | awk '{print $1}' )
```
Comment (msalle) [inline travis-ci/make_rpms.sh]: this looks quite ugly and would actually rebuild the $rpmpath. The output of something like `rpm --requires -qp "$rpmpath" | cut -d' ' -f1`. That's still ignoring version requirements, so still not ideal, but looks a lot cleaner.
Maybe I'm missing something...?

File: travis-ci/run_task_inside_docker.sh [hunk 4]
```diff
@@ -8,59 +8,25 @@ TASK=$2
...
-#    # for globus-gram-audit:
-#    packages+=('perl(DBI)')
-#    # for globus-scheduler-event-generator:
-#    packages+=(redhat-lsb-core)
-#    # for myproxy-oauth
-#    packages+=(m2crypto mod_ssl mod_wsgi pyOpenSSL python-crypto)
-#    # for gsi-openssh
-#    packages+=(pam libedit libedit-devel)
-#fi
+# Standard packages neded for all tasks
```
Comment (msalle) [inline travis-ci/run_task_inside_docker.sh]: small typo `neded` -> `needed`

**Review by msalle — COMMENTED** [review]
I've left a view comments on things that I find a bit unclear. If it all works it is fine for me for the time being, but probably good to look at, at some point.

File: packaging/fedora/uberftp.spec (hunk 1 above)
Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: The original file name (and URL) was meant to build from the source tarball as provided by GitHub (i.e. the untampered contents of the Git repo without the `.git` dir). For the RPM builds on Travis-CI the actual dist tarball is created prior and used instead. And due to the current configuration in `configure.ac`:
```
[...]
AC_INIT(UberFTP (2.9 Beta), [2.9-Beta], [Grid Community Forum <discuss@gridcf.org>],[uberftp-client])
[...]
```
...`make dist` creates a tarball named `uberftp-client-2.9-Beta.tar.gz`. The URL in the RPM spec is ignored by `rpm-build` as long as the source file name matches IIUIC.

This needs some rethinking about what we want to have in the RPM spec file. I think, as soon as we upload the dist tarballs to `repo.gridcf.org` (would be my plan for the next PR, but needs help from OSG people with the encrypted credentials to be used by the GitHub action) we should just use a URL pointing there. For the naming we should include a command line in the RPM spec that determines the relevant name parts from the `configure.ac`. So on version changes we only need to change things there then.

File: packaging/fedora/uberftp.spec (hunk 2 above)
Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: See above.

File: travis-ci/make_rpms.sh (hunk 3 above)
Comment (fscheiner) [reply travis-ci/make_rpms.sh]: I'm happy with anything better than what's in the PR. The actual problem is described in https://github.com/gridcf/UberFTP/pull/16#issuecomment-974179081. No build dependencies are installed prior to the build of the RPMs not even in [the original file](https://github.com/gridcf/gct/blob/5165cc9685d5dd5a40df445ca671f06771da8099/travis-ci/make_rpms.sh) - maybe this was never tested and the original deploy step with Travis-CI created SRPMs only, which would work despite the missing build dependencies.

File: travis-ci/run_task_inside_docker.sh (hunk 4 above)
Comment (fscheiner) [reply travis-ci/run_task_inside_docker.sh]: Thanks, I'll address this after the above conversations were resolved.

File: packaging/fedora/uberftp.spec (hunk 1 above)
Comment (msalle) [reply packaging/fedora/uberftp.spec]: I think it is best to rename the `uberftp-client` in the `configure.ac` into uberftp as that's how we have it in EPEL and I think dist tarballs have never been released.
Then I would put the name, the version and release manually one time in the spec file (something like [here](https://src.fedoraproject.org/rpms/curl/blob/rawhide/f/curl.spec#_2-4) and derive all other occurrences in the spec file from there. You could use magic to get it from the configure.ac but I'm not sure that's worth the effort.

File: travis-ci/make_rpms.sh (hunk 3 above)
Comment (msalle) [reply travis-ci/make_rpms.sh]: Right, so probably
```
sudo yum install -y $(rpm --requires -qp "$rpmpath" | cut -d' ' -f1)
```
is already better, except that it won't take into account requirements with version constraints.
Have you actually tried
```
sudo yum-builddep -y --releasever=... $rpmpath
```
I didn't succeed without the releasever set to e.g. 7, but with it it seems fine

File: packaging/fedora/uberftp.spec (hunk 1 above)
Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: Sounds good! I'll take this into account and push it in an additional change.

Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: About the version number. I agree, that it's best to hardcode it in the spec file, as deriving it from the `configure.ac` might not even work during runtime assuming that the sources aren't yet extracted when `rpm-build` starts processing the spec file. Late yesterday I thought about a `VERSION` file in the UberFTP sources and use `m4` to get it into the configure.ac during runtime (e.g. [this description](https://newbedev.com/read-a-version-number-from-a-file-in-configure-ac) looks like a good option, especially the variant which auto-includes  the `VERSION` file into the `Makefile`). Then we would need to change it "only" in the `VERSION` file and the RPM spec for releases.

Comment (msalle) [reply packaging/fedora/uberftp.spec]: In which case I would definitely go for the `m4_normalize(m4_include([VERSION]))` way of doing it.

File: travis-ci/make_rpms.sh (hunk 3 above)
Comment (fscheiner) [reply travis-ci/make_rpms.sh]: I'll give that a try.

****

_Sorry, apparently I didn't hit the "Comment" button until now._ :-/

**Comment by matyasselmeci** [comment]
`remove_all_gct_rpms()` shouldn't be in there for the UberFTP build. It was in the GCT build to make sure we were only using the GCT binaries that we compiled, but that doesn't apply to UberFTP.

**Comment by gerrith3** [comment]
> @gerrith3
> 
> > I will take a look but getting an entire distro supported can be difficult aka expensive. It may depend on where they are pulling their sources from - if they are using fedora or equivalent build environments that are copied into RockyLinux or AlmaLinux there may be a chance.
> 
> Great to hear that, I assume both RockyLinux and AlmaLinux are using the official sources of RHEL8 (which supports ppc64le), i.e. like CentOS8 did until Red Hat changed that to CentOS Stream 8.

Hey @fscheiner I forgot to update you here but both RockyLinux and AlmaLinux are using our Power servers at OSU's OSL for testing, so both should have Power LE support (ppc64le).

**Comment by fscheiner** [comment]
@matyasselmeci 
> `remove_all_gct_rpms()` shouldn't be in there for the UberFTP build. It was in the GCT build to make sure we were only using the GCT binaries that we compiled, but that doesn't apply to UberFTP.

Thank you for the clarification. I had some thoughts about that and believe I understand it better now but correct me if I'm wrong: It works for the GCT, because - after all Globus related packages are removed - the build starts from the beginning (i.e. with `globus-common` which does not depend on any `globus-*` packages) and works through the build-dependency ordered list of packages ([`packaging/fedora/ORDERING`](https://github.com/gridcf/gct/blob/c302831b49a105c968742631e7b808cbff6f320b/packaging/fedora/ORDERING)) and builds **and** installs all RPMs in a single call to `build_all_for_target()` (from [`make_rpms.sh`](https://github.com/gridcf/gct/blob/5165cc9685d5dd5a40df445ca671f06771da8099/travis-ci/make_rpms.sh).

**Comment by fscheiner** [comment]
@gerrith3
> Hey @fscheiner I forgot to update you here but both RockyLinux and AlmaLinux are using our Power servers at OSU's OSL for testing, so both should have Power LE support (ppc64le).

Thanks for the update, that's good news. Unfortunately both don't yet provide Docker images for ppc64le over at https://hub.docker.com/r/rockylinux/rockylinux/tags and https://hub.docker.com/_/almalinux?tab=tags but only for amd64 and arm64. Well, they currently also only offer installation disc images for amd64 and arm64 (https://rockylinux.org/download, https://mirrors.almalinux.org/isos.html) as I just found out, so I assume Docker images for ppc64le will eventually come.

File: travis-ci/run_task_inside_docker.sh (hunk 4 above)
Comment (fscheiner) [reply travis-ci/run_task_inside_docker.sh]: Adressed with https://github.com/gridcf/UberFTP/pull/16/commits/3b29e7b5bc194a9d2183ccbea5b943123b31db92

File: packaging/fedora/uberftp.spec (hunk 1 above)
Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: Addressed with https://github.com/gridcf/UberFTP/pull/16/commits/59bdc2b0cd195516a7a8187e295ae9ce555f19fe

File: travis-ci/make_rpms.sh (hunk 3 above)
Comment (fscheiner) [reply travis-ci/make_rpms.sh]: No longer needed with https://github.com/gridcf/UberFTP/pull/16/commits/bbaf36959599cf768db97e5cf189909e77ee01f2

File: packaging/fedora/uberftp.spec (hunk 2 above)
Comment (fscheiner) [reply packaging/fedora/uberftp.spec]: Fixed with https://github.com/gridcf/UberFTP/pull/16/commits/59bdc2b0cd195516a7a8187e295ae9ce555f19fe

**Comment by fscheiner** [comment]
@msalle:
My latest changes should address the comments from [your last review](https://github.com/gridcf/UberFTP/pull/16#pullrequestreview-816158696).

Successful build here: https://app.travis-ci.com/github/fscheiner/UberFTP/builds/242806697

The addition of a GitHub Action/Worklfow for build tests and uploading source tarballs will be implemented in another PR.

So if there are no objections I'll merge these changes _as is_ tomorrow.

**Pull request merged**